File: sandbox/BitStringRank.py
"""
Defines a class that performs constant-time
ranking on bit strings, as well as some tests.

IMPORTANT:
This implementation is for algorithm explanation
only, it is not for production use, due to
it's memory consumption. It is very important
for a bit string to be stored as efficiently as possible,
which this implementation dues not guarantee.
"""
import numpy as np
import logging

log = logging.getLogger(__name__)


class BitStringRank(object):
    """
    Class  that performs constant-time
    ranking on bit strings.

    A  bit-string 'rank' for a given index
    is the number of bits set to {0, 1} up until
    and including the index.
    """

    def __init__(self, string, bucket_size):
        """
        Creates the BitStringRank object.

        :param string: An iterable consisting of 0s and 1s,
            can be numeric or text.

        :param bucket_size: Bucket size.
        """
        super(BitStringRank, self).__init__()
        assert(bucket_size > 1)

        self.string = [bool(c) for c in string]
        self.n = len(string)
        self.bucket_size = bucket_size
        superbucket_size = bucket_size ** 2

        #   create the buckets
        self.buckets = np.zeros(self.n / bucket_size + 1, np.uint8)
        self.superbuckets = np.zeros(self.n / bucket_size ** 2 + 1, np.int)

        #   populate the buckets with values
        current_sum = 0
        #   iterate through the string
        for ind, value in enumerate(self.string):

            #   increment current sum
            current_sum += value

            #   when at the last element of the current bucket
            #   update the next bucket's size
            if (ind % bucket_size == (bucket_size - 1)):

                #   index of the next bucket
                bucket_ind = ind / self.bucket_size + 1
                #   index of the superbucket the next bucket belongs to
                superbucket_ind = bucket_ind / self.bucket_size

                #   update superbucket when crossing
                #   the superbucket boundary
                if ind % superbucket_size == (superbucket_size - 1):
                    self.superbuckets[superbucket_ind] = current_sum

                #   update the bucket relative to it's superbucket
                self.buckets[bucket_ind] = current_sum - \
                    self.superbuckets[superbucket_ind]

    def rank(self, i):
        """
        Returns the number of bits set (having the value 1)
        in the first i bits of this bit string.

        :param i: See above.
        """
        assert (i >= 0) & (i <= self.n)

        bucket_ind = i / self.bucket_size
        superbucket_ind = bucket_ind / self.bucket_size
        return self.superbuckets[superbucket_ind] + \
            self.buckets[bucket_ind] + \
            sum(self.string[bucket_ind * self.bucket_size: i])

File: src/BitStringRank.cpp
#include "../include/BitStringRank.h"

BitStringRank::BitStringRank(const BitVector& bit_vector, UInt bucket_size)
    : _bucket_size(bucket_size), _bit_vector(bit_vector) {

    // bucket_size < 2^(BucketType bit number)
    assert(bucket_size < (1 << (sizeof(BucketType) * 8)));

    this->_length = bit_vector.size();
    //std::cout << "-- Creatin new BitStringRank" << std::endl;
    //std::cout << "Bit vector has size of " << _length << std::endl;
    
    // fill with zeros
    this->_buckets.resize(_length / _bucket_size + 1, 0);
    this->_superbuckets.resize(_length / (_bucket_size * _bucket_size) + 1, 0);
    //std::cout << "BS, SBS " << _buckets.size() << " " << _superbuckets.size() << std::endl;

    // populate buckets with values
    UInt current_sum = 0;
    UInt index = 0;
    UInt superbucket_size = _bucket_size * _bucket_size;
    BitVector::const_iterator bit_vec_it = this->_bit_vector.begin();
    for ( ; bit_vec_it != this->_bit_vector.end(); bit_vec_it++, index++) {

        current_sum += *bit_vec_it;
        
        // if we are exiting a bucket, store the result
        if (index % _bucket_size == (_bucket_size - 1)) {

            // indexes of bucket and superbucket for current index
            UInt bucket_index = index / _bucket_size + 1;
            UInt superbucket_index = bucket_index / _bucket_size;

            // if we are exiting superbucket, also update that
            if (index % superbucket_size == (superbucket_size - 1)) {
                _superbuckets[superbucket_index] = current_sum;
            }

            // update the bucket relative to the current superbucket
            this->_buckets[bucket_index] = current_sum -
                this->_superbuckets[superbucket_index];
        }
        
    }

    /*
    std::cout << "Superbucksets" << std::endl;
    for (std::size_t i = 0; i < _superbuckets.size(); i++)
        std::cout << _superbuckets[i] << " ";
    std::cout << std::endl;
    
    std::cout << "Buckets" << std::endl;
    for (std::size_t i = 0; i < _buckets.size(); i++)
        std::cout << _buckets[i] << " ";
    std::cout << std::endl;
    */
}

UInt BitStringRank::rank(UInt i) {
    //std::cout << "BitStringRank rank " << i << ", length " << this->_length << std::endl;
    assert(i <= this->_length);

    if (i == 0) return 0;

    UInt bucket_index = i / this->_bucket_size;
    UInt superbucket_index = bucket_index / this->_bucket_size;

    UInt startIndex = bucket_index * this->_bucket_size;
    UInt sum = 0;
    for (std::size_t ind = startIndex; ind < i; ind++)
    {
        sum += this->_bit_vector[ind];
    }

    return sum + this->_buckets[bucket_index] + 
        this->_superbuckets[superbucket_index];
}
